File: Haskell/Implementations/Command/Email.hs
module Email
    ( Email(..)
    , emptyEmail
    , signEmail
    , sendEmail
    ) where

import Data.List (intercalate)

data Email = MkEmail
    { subject   :: String
    , body      :: String
    , signature :: Maybe String
    }

emptyEmail :: Email
emptyEmail = MkEmail { subject = "", body = "", signature = Nothing }

sendEmail :: Email -> String -> IO ()
sendEmail email recipient = do
    putStrLn $ intercalate
        "\n"
        (  [ "+ New email"
           , "| Recipient: " <> recipient
           , "| Subject: " <> subject email
           ]
        <> (case signature email of
               Nothing  -> []
               Just sig -> ["| Signature: " <> sig]
           )
        <> ["| Body: " <> body email]
        )

signEmail :: Email -> Email
signEmail email = email { signature = Just "Signed by ConsoleMailServer" }

File: Java/Implementations/Command/Email.java
import java.util.function.Function;

public class Email {
	public final String subject;
	public final String body;
	public final String signature;

	public Email(String subject, String body, String signature) {
		this.subject = subject;
		this.body = body;
		this.signature = signature;
	}

	public static class Builder implements EmailEditor {
		private String subject = "";
		private String body = "";
		private String signature;

		public EmailEditor subject(String subject) {
			this.subject = subject;
			return this;
		}

		public EmailEditor body(String body) {
			this.body = body;
			return this;
		}

		public EmailEditor sign(String signature) {
			this.signature = signature;
			return this;
		}

		public Email create() {
			return new Email(this.subject, this.body, this.signature);
		}
	}
}
